File: node-advisor-core/test/routes.integration.test.ts
import { test } from "node:test"

import { AdvisorCore } from "../src/AdvisorCore"
import { StorageDownloadPayload } from "../src/payloads"
import {
  assertBufferSuccess,
  assertJsonSuccess,
  assertStreamSuccess,
  createAdvisor,
  createPayloads,
  resolveStorageDownloadPayload,
} from "./integration.helpers"

const payloads = createPayloads()
const hasAdvisorToken = Boolean(process.env.ADVISOR_TOKEN)

let advisor: AdvisorCore | undefined
let storageDownloadPayloadPromise: Promise<StorageDownloadPayload> | undefined

function getAdvisor(): AdvisorCore {
  if (!advisor) {
    advisor = createAdvisor()
  }

  return advisor
}

function getStorageDownloadPayload(): Promise<StorageDownloadPayload> {
  if (!storageDownloadPayloadPromise) {
    storageDownloadPayloadPromise = resolveStorageDownloadPayload(
      getAdvisor(),
      payloads.storageListPayload,
    )
  }

  return storageDownloadPayloadPromise
}

if (!hasAdvisorToken) {
  test("integration setup requires ADVISOR_TOKEN", () => {
    throw new Error("Set ADVISOR_TOKEN before running the Node integration tests.")
  })
} else {
  for (const methodName of ["getDaily", "getHourly", "getPeriod"] as const) {
    test(`forecast.${methodName}`, async () => {
      const response = await getAdvisor().forecast[methodName](payloads.weatherPayload)
      assertJsonSuccess(response)
    })
  }

  for (const methodName of ["getDaily", "getHourly", "getPeriod"] as const) {
    test(`observed.${methodName}`, async () => {
      const response = await getAdvisor().observed[methodName](payloads.weatherPayload)
      assertJsonSuccess(response)
    })
  }

  test("observed.getStationData", async () => {
    const response = await getAdvisor().observed.getStationData(payloads.stationPayload)
    assertJsonSuccess(response)
  })

  for (const methodName of ["getFireFocus", "getLightning"] as const) {
    test(`observed.${methodName}`, async () => {
      const response = await getAdvisor().observed[methodName](payloads.radiusPayload)
      assertJsonSuccess(response)
    })
  }

  for (const methodName of ["getFireFocusByGeometry", "getLightningByGeometry"] as const) {
    test(`observed.${methodName}`, async () => {
      const response = await getAdvisor().observed[methodName](payloads.geometryPayload)
      assertJsonSuccess(response)
    })
  }

  test("observed.getLightningDetails", async () => {
    const response = await getAdvisor().observed.getLightningDetails(payloads.lightningDetailsPayload)
    assertJsonSuccess(response)
  })

  test("observed.getLightningLite", async () => {
    const response = await getAdvisor().observed.getLightningLite(payloads.lightningLitePayload)
    assertJsonSuccess(response)
  })

  test("currentWeather.get", async () => {
    const response = await getAdvisor().currentWeather.get(payloads.currentWeatherPayload)
    assertJsonSuccess(response)
  })

  for (const methodName of ["getDaily", "getMonthly"] as const) {
    test(`climatology.${methodName}`, async () => {
      const response = await getAdvisor().climatology[methodName](payloads.climatologyPayload)
      assertJsonSuccess(response)
    })
  }

  test("monitoring.getAlerts", async () => {
    const response = await getAdvisor().monitoring.getAlerts()
    assertJsonSuccess(response)
  })

  test("stations.getLastData", async () => {
    const response = await getAdvisor().stations.getLastData(payloads.stationsLastDataPayload)
    assertJsonSuccess(response)
  })

  test("plan.getInfo", async () => {
    const response = await getAdvisor().plan.getInfo(payloads.planInfoPayload)
    assertJsonSuccess(response)
  })

  test("plan.getRequestDetails", async () => {
    const response = await getAdvisor().plan.getRequestDetails(payloads.requestDetailsPayload)
    assertJsonSuccess(response)
  })

  test("plan.getLocale", async () => {
    const response = await getAdvisor().plan.getLocale(payloads.planLocalePayload)
    assertJsonSuccess(response)
  })

  for (const methodName of [
    "getForecastDaily",
    "getForecastHourly",
    "getObservedDaily",
    "getObservedHourly",
  ] as const) {
    test(`chart.${methodName}`, async () => {
      const response = await getAdvisor().chart[methodName](payloads.weatherChartPayload)
      assertBufferSuccess(response)
    })
  }

  test("storage.listFiles", async () => {
    const response = await getAdvisor().storage.listFiles(payloads.storageListPayload)
    assertJsonSuccess(response)
  })

  test("storage.downloadFile", async () => {
    const response = await getAdvisor().storage.downloadFile(await getStorageDownloadPayload())
    assertBufferSuccess(response)
  })

  test("storage.downloadFileByStream", async () => {
    const response = await getAdvisor().storage.downloadFileByStream(await getStorageDownloadPayload())
    await assertStreamSuccess(response)
  })

  test("staticMap.get", async () => {
    const response = await getAdvisor().staticMap.get(payloads.staticMapPayload)
    assertBufferSuccess(response)
  })

  test("tms.get", async () => {
    const response = await getAdvisor().tms.get(payloads.tmsPayload)
    assertBufferSuccess(response)
  })

  test("pmtiles.get", async () => {
    const response = await getAdvisor().pmtiles.get(payloads.pmtilesPayload)
    assertBufferSuccess(response)
  })

  test("schema.getDefinition", async () => {
    const response = await getAdvisor().schema.getDefinition()
    assertJsonSuccess(response)
  })

  test("schema.postDefinition", async () => {
    const response = await getAdvisor().schema.postDefinition(payloads.schemaDefinitionPayload)
    assertJsonSuccess(response)
  })
}

File: go-advisor-core/tests/routes-integration_test.go
//go:build integration

package test

import (
	"io"
	"testing"
)

func TestForecastRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	routes := map[string]func(WeatherPayload) (AdvisorResponse, error){
		"GetDaily":  advisor.Forecast.GetDaily,
		"GetHourly": advisor.Forecast.GetHourly,
		"GetPeriod": advisor.Forecast.GetPeriod,
	}

	for name, route := range routes {
		t.Run(name, func(t *testing.T) {
			response, err := route(payloads.weatherPayload)
			assertJSONSuccess(t, response, err)
		})
	}
}

func TestObservedWeatherRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	routes := map[string]func(WeatherPayload) (AdvisorResponse, error){
		"GetDaily":  advisor.Observed.GetDaily,
		"GetHourly": advisor.Observed.GetHourly,
		"GetPeriod": advisor.Observed.GetPeriod,
	}

	for name, route := range routes {
		t.Run(name, func(t *testing.T) {
			response, err := route(payloads.weatherPayload)
			assertJSONSuccess(t, response, err)
		})
	}
}

func TestObservedStationData(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	response, err := advisor.Observed.GetStationData(payloads.stationPayload)
	assertJSONSuccess(t, response, err)
}

func TestObservedRadiusRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	routes := map[string]func(RadiusPayload) (AdvisorResponse, error){
		"GetFireFocus":        advisor.Observed.GetFireFocus,
		"GetLightning":        advisor.Observed.GetLightning,
		"GetLightningDetails": advisor.Observed.GetLightningDetails,
	}

	for name, route := range routes {
		t.Run(name, func(t *testing.T) {
			payload := payloads.radiusPayload
			if name == "GetLightningDetails" {
				payload = payloads.lightningDetailsPayload
			}

			response, err := route(payload)
			assertJSONSuccess(t, response, err)
		})
	}
}

func TestObservedGeometryRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	routes := map[string]func(GeometryPayload) (AdvisorResponse, error){
		"GetFireFocusByGeometry": advisor.Observed.GetFireFocusByGeometry,
		"GetLightningByGeometry": advisor.Observed.GetLightningByGeometry,
	}

	for name, route := range routes {
		t.Run(name, func(t *testing.T) {
			response, err := route(payloads.geometryPayload)
			assertJSONSuccess(t, response, err)
		})
	}
}

func TestObservedLightningLite(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	response, err := advisor.Observed.GetLightningLite(payloads.lightningLitePayload)
	assertJSONSuccess(t, response, err)
}

func TestCurrentWeather(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	response, err := advisor.CurrentWeather.Get(payloads.currentWeatherPayload)
	assertJSONSuccess(t, response, err)
}

func TestClimatologyRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	routes := map[string]func(ClimatologyPayload) (AdvisorResponse, error){
		"GetDaily":   advisor.Climatology.GetDaily,
		"GetMonthly": advisor.Climatology.GetMonthly,
	}

	for name, route := range routes {
		t.Run(name, func(t *testing.T) {
			response, err := route(payloads.climatologyPayload)
			assertJSONSuccess(t, response, err)
		})
	}
}

func TestMonitoringAlerts(t *testing.T) {
	advisor := newIntegrationAdvisor(t)

	response, err := advisor.Monitoring.GetAlerts()
	assertJSONSuccess(t, response, err)
}

func TestStationsLastData(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	response, err := advisor.Stations.GetLastData(payloads.stationsLastDataPayload)
	assertJSONSuccess(t, response, err)
}

func TestPlanRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	t.Run("GetInfo", func(t *testing.T) {
		response, err := advisor.Plan.GetInfo(payloads.planInfoPayload)
		assertJSONSuccess(t, response, err)
	})

	t.Run("GetRequestDetails", func(t *testing.T) {
		response, err := advisor.Plan.GetRequestDetails(payloads.requestDetailsPayload)
		assertJSONSuccess(t, response, err)
	})

	t.Run("GetLocale", func(t *testing.T) {
		response, err := advisor.Plan.GetLocale(payloads.planLocalePayload)
		assertJSONSuccess(t, response, err)
	})
}

func TestChartRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	routes := map[string]func(WeatherPayload) (io.ReadCloser, error){
		"GetForecastDaily":  advisor.Chart.GetForecastDaily,
		"GetForecastHourly": advisor.Chart.GetForecastHourly,
		"GetObservedDaily":  advisor.Chart.GetObservedDaily,
		"GetObservedHourly": advisor.Chart.GetObservedHourly,
	}

	for name, route := range routes {
		t.Run(name, func(t *testing.T) {
			reader, err := route(payloads.weatherChartPayload)
			assertBinarySuccess(t, reader, err)
		})
	}
}

func TestStorageRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()
	downloadPayload := resolveStorageDownloadPayload(t)

	t.Run("ListFiles", func(t *testing.T) {
		response, err := advisor.Storage.ListFiles(payloads.storageListPayload)
		assertJSONSuccess(t, response, err)
	})

	t.Run("DownloadFile", func(t *testing.T) {
		reader, err := advisor.Storage.DownloadFile(downloadPayload)
		assertBinarySuccess(t, reader, err)
	})
}

func TestStaticMap(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	reader, err := advisor.StaticMap.Get(payloads.staticMapPayload)
	assertBinarySuccess(t, reader, err)
}

func TestTms(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	reader, err := advisor.Tms.Get(payloads.tmsPayload)
	assertBinarySuccess(t, reader, err)
}

func TestPmtiles(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	reader, err := advisor.Pmtiles.Get(payloads.pmtilesPayload)
	assertBinarySuccess(t, reader, err)
}

func TestSchemaRoutes(t *testing.T) {
	advisor := newIntegrationAdvisor(t)
	payloads := createIntegrationPayloads()

	t.Run("GetDefinition", func(t *testing.T) {
		response, err := advisor.Schema.GetDefinition()
		assertJSONSuccess(t, response, err)
	})

	t.Run("PostDefinition", func(t *testing.T) {
		response, err := advisor.Schema.PostDefinition(payloads.schemaDefinitionPayload)
		assertJSONSuccess(t, response, err)
	})
}
